packages/protocol/contracts/layer1/core/impl/Inbox.sol

```diff
@@ -245,6 +246,98 @@ contract Inbox is IInbox, IForcedInclusionStore, EssentialContract {
     }
 
+    /// @dev Verifies a batch proof covering multiple consecutive proposals.
+    ///
+    /// The proof must cover a contiguous range of proposals that includes (or starts after) the
+    /// last finalized proposal. The transitionHashs array contains N+1 hashes for N proposals:
+    /// - transitionHashs[0] is the parent state (before first proved proposal)
+    /// - transitionHashs[N] is the final state (after last proved proposal)
+    ///
+    /// Example: Proving proposals 4-7 when lastFinalizedProposalId=4
+    ///
+    ///       lastFinalizedProposalId                nextProposalId
+    ///                             ↓                             ↓
+    ///     0     1     2     3     4     5     6     7     8     9
+    ///     ■─────■─────■─────■─────■─────□─────□─────□─────□─────
+    ///                 ↑     └──── proof coverage ────┘
+    ///                 firstProposalParentId
+    ///
+    ///     transitionHashs = [hash3, hash4, hash5, hash6, hash7]
+    ///                         ↑                           ↑
+    ///    state.lastFinalizedProposalHash             new state.lastFinalizedProposalHash
+    ///
+    /// Key validation rules:
+    /// 1. The proof must start at or before the last finalized proposal
+    /// 2. The proof must advance finalization (lastProposalId > lastFinalizedProposalId)
+    /// 3. The proof cannot extend beyond proposed blocks (lastProposalId < nextProposalId)
+    /// 4. The transition hash at lastFinalizedProposalId must match the stored hash
+    ///
+    /// Note: Since each proposal contains its parent hash, verifying only the last proposal's hash
+    /// cryptographically commits to the entire chain of proposals.
+    function prove2(bytes calldata _data, bytes calldata _proof) external nonReentrant {
+        unchecked {
+            CoreState memory state = _state;
+            ProveInput2 memory input = abi.decode(_data, (ProveInput2));
+            // Note: ProveInput2 is defined in IInbox
+
+            // transitionHashs has N+1 elements for N proposals (parent hash + N transition hashes)
+            require(input.transitionHashs.length > 1, InsufficientTransitionHashes());
+
+            uint48 numProvedProposals = uint48(input.transitionHashs.length - 1);
+            uint256 firstProposalParentId = input.lastProposalId - numProvedProposals;
+
+            // Ensure the proof range is valid:
+            // - Must start at or before the last finalized proposal (to include its transition hash)
+            // - Must advance past the current finalization point
+            // - Cannot prove proposals that haven't been proposed yet
+            require(
+                firstProposalParentId <= state.lastFinalizedProposalId, ProofRangeStartTooLate()
+            );
+            require(
+                input.lastProposalId >= state.lastFinalizedProposalId + _minProposalsToFinalize,
+                InsufficientProposalsToFinalize()
+            );
```

Comment:
Is there a chance that, a user only wants to prove and finalize his / her own proposals? Like the current `lastFinalizedProposalId`  is `100`, Alice proposed `101`, and Bob proposed `102`, but Alice only wants to submit the proof of her own `101` in time ASAP and doesn't want to generate the proofs for Bob?

Reply:
then the chain will finalize very slowly and eventual the ring buffer will be exhausted. we have to enforce each proof to prove enough proposals so finalization can catch up with new proposals.